### DISCUSSION TITLE
importDB and importInto

### DISCUSSION THREAD

**Discussion by laukaichung** [dexie/Dexie.js#2063] [Q&A] [answered]
Hello,

I used `importInto` to merge the file with the current database which is great, but it won't trigger upgraders if the file version is outdated and even with the `acceptVersionDiff: true` option.

I just found this [thread](https://github.com/dexie/Dexie.js/issues/1337) and it helped me tackle the situation. But I can't tell how it works and the difference between `importDB ` and `importInto`. 

Does `importDB` create a separate Dexie instance from the imported file which is independent of the current database?
But how could it magically replace the current database and run the required upgraders after running?

````
tempDb.close();
await db.open(); 
// I would think there would be something like `await db.import(tempDB) , but it just works.
````

**Comment by dfahlander** [comment, marked as answer]
There is no upgrading support built into dexie-export-import. It's just that if you use importDB it will create a database with the same version as it was when it was exported. Opening it again using a declared Dexie instance with a higher version and upgraders declared, these upgraders will run into the data.

In your case, you have already existing data and you want to keep it and just merge the import into the existing database. In this case you might have to implement migration your custom way for the imported data.

1. use [peakImportFile()](https://dexie.org/docs/ExportImport/dexie-export-import#usage) to find out which data version the import file is on.
    ```ts
    const { formatName, formatVersion, data }  = await peakImportFile(blob);
    if (formatName !== "dexie") throw new Error(`The blob does not contain a dexie export`);
    const dataVersion = data.databaseVersion; // Base your migration code on this
    ```

2. Use the [transform option](https://dexie.org/docs/ExportImport/dexie-export-import#staticimportoptions-and-importoptions) to upgrade data row-by-row while it is being imported.

    ```ts
    const transform = (table, value, key?) => {
      if (dataVersion < 3) {
        // Upgrade value (which is one single row) before import
        const names = value.name.split(' ');
        value.fistName = names.shift();
        value.lastName = names.pop();
        value.middleNames = names;
        delete value.name;
      }
      return { value, key };
    };
    await importInto(db, blob, { transform });
    ```

**Comment by laukaichung** [comment]
Thank you for the explanation. I will come back to this thread when I run into issues. Really appreciate your efforts.
